seqr/views/utils/dataset_utils.py

```diff
@@ -421,8 +421,7 @@ def _load_rna_seq(model_cls, file_path, user, mapping_file, ignore_extra_samples
         samples = validate_samples(samples, sample_id_to_tissue_type, warnings)
 
     # Delete old data
-    individual_db_ids = {s.individual_id for s in samples}
-    to_delete = model_cls.objects.filter(sample__individual_id__in=individual_db_ids).exclude(sample__data_source=data_source)
+    to_delete = model_cls.objects.filter(sample__in=samples).exclude(sample__data_source=data_source)
     prev_loaded_individual_ids = set(to_delete.values_list('sample__individual_id', flat=True))
```

Comment:
Please also review the logic for the code below using `prev_loaded_individual_ids`.

Reply:
this is not the behavior we want- we want to remove all data previously loaded for this individual and tissue type. If this data is from a new file it will have a new sample model and then these won't be deleted anymore even if its the same tissue type

Reply:
The `samples` meet the per `individual and tissue type` conditions. So the new code removes all data previously loaded for the individual and tissue type.

Reply:
can you add unit tests for this behavior? We should test that data for an individual is not removed when its a new tissue type, but is removed for an existing tissue type when reloaded from a new file

Reply:
Yes, I can add some tests. And I also need to make some changes to allow TPM data to have more than one tissue type. The existing code doesn't create new sample models for new tissues but validates the tissue and skips loading the unmatched tissue data.

Reply:
After adding a sample of `fibroblasts` tissue and RnaSeqTpm data for the sample, the test for loaded data failed for both Outlier and Tpm.
The response for the Outlier is:
```
{'info': ['Parsed 1 RNA-seq samples', 'Attempted data loading for 1 RNA-seq samples in the following 1 projects: 1kg project nåme with uniçøde'],
'sampleGuids': ['S000152_na19675_d2'],
'warnings': ['Skipped loading for 1 samples already loaded from this file'],
'fileName': 'rna_sample_data__outlier__2020-04-15T00:00:00.json.gz' }
```
For TPM it is:
```
{'info': ['Parsed 1 RNA-seq samples', 'Attempted data loading for 0 RNA-seq samples in the following 0 projects: '], 
'sampleGuids': [],
'warnings': ['Skipped data loading for the following 1 samples due to mismatched tissue type: NA19675_D2 (muscle to fibroblasts)', 'Skipped loading for 1 samples already loaded from this file']',
fileName': 'rna_sample_data__tpm__2020-04-15T00:00:00.json.gz'}
```
While the expected response is:
```
{'info': ['Parsed 1 RNA-seq samples', 'Attempted data loading for 0 RNA-seq samples in the following 0 projects: '], 'warnings': ['Skipped loading for 1 samples already loaded from this file'],
'sampleGuids': [], 'fileName': <ANY>}
```

Reply:
Make sure you are using a different file name for the new tissue type in your tests. If you are still seeing an issue, that either means you wrote the test case wrong or it means you did not correctly update the application behavior

Reply:
The new sample has a new tissue type and data source, as shown follows.
```
        "tissue_type": "F",
        "data_source": "fibroblasts_samples.tsv.gz",
```

Reply:
The Outlier test fails because the function `_find_matching_sample_records` return more samples than expected. It returns all the samples with the same `sample_id` and different tissue types. It results in duplicating data for each tissue type.
The TMP test doesn't load the loaded data which is correct because it filters out the extra samples by matching the tissue type. The Outlier can't do the filtering because Outlier data don't have a tissue type.
The TMP tests fails for it has an extra warning message while filtering out the unwanted samples. And it also deletes the existing TMP data unexpectedly if we don't change the code for the `to_delete`.

To solve all these issues, I'm going to add tissue type filtering for both Outlier and TPM samples and remove the warnings if samples are filtered out.
But there is still an issue: The Outlier samples must have a `None` tissue type. It could have backward compatibility issue. There are quite a lot of Outlier data share the sample models with TPM data. These sample models' tissue type is not `None`. When loading new Outlier data the existing Outlier data won't be deleted automatically.

Reply:
> The Outlier samples must have a None tissue type.

They don't have to have a None tossue type but they can. If we load Outlier data first and don't have a tissue type, and then later add TPM data and do have a tissue type, we want to set the tissue type on the existing sample model, not create a new model with no tissue type

> When loading new Outlier data the existing Outlier data won't be deleted automatically.

You will have to add special handling for if you are loading new data with no known tissue type, you want to delete any data previously loaded on any tissue type in that case

Reply:
Alternatively, we could stop sharing sample models between different calling types. So there would no longer be the case where Outlier and TPM models share the same sample, each would have their own sample. That would also solve a bunch of the issue you outline here